### ISSUE TITLE
stringReplaceTokens

### ISSUE THREAD

**Issue by Kezarus-tech** [Alphish/gm-community-toolbox#92]
Description below.

```
//Receives a text with tokens between '<>' and an object with variables
//and substitute all found tokens in the text and returns it
function stringReplaceTokens(wText, wObject){
	if( !is_string(wText) || !is_struct(wObject) ){
		return "";
	}
	
	var result = wText;
	var arrVariableNames = variable_struct_get_names(wObject);
	for(var i=0; i<array_length(arrVariableNames); i++;){
		var variableName = arrVariableNames[i];
		var variableToken = "<"+arrVariableNames[i]+">";
		var variableValue = struct_get_value(wObject, variableName, "");
		
		result = string_replace_all(result, variableToken, variableValue);
	}
	
	return result;
}


function stringReplaceTokens_test(){
	var textInput, textResult, textExpected, wObject;
	
	//TEST 1
	textInput = 5; //NOT A TEXT
	wObject = {};
	textExpected = "";
	var textResult = stringReplaceTokens(textInput, wObject);
	assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 1");
	
	//TEST 2
	textInput = "Simple text";
	wObject = 5; //NOT AN OBJECT
	textExpected = "";
	var textResult = stringReplaceTokens(textInput, wObject);
	assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 2");
	
	// TEST 3: No tokens in text
    textInput = "Simple text without tokens";
    wObject = {name: "John"};
    textExpected = "Simple text without tokens";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 3");

    // TEST 4: Single token replacement
    textInput = "Hello, <name>!";
    wObject = {name: "John"};
    textExpected = "Hello, John!";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 4");

    // TEST 5: Multiple tokens replacement
    textInput = "Hello, <name>! Your age is <age>.";
    wObject = {name: "Alice", age: "30"};
    textExpected = "Hello, Alice! Your age is 30.";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 5");

    // TEST 6: Non-existent tokens in text
    textInput = "Hello, <name>! Your age is <age>. Your city is <city>.";
    wObject = {name: "Alice", age: "30"};
    textExpected = "Hello, Alice! Your age is 30. Your city is <city>.";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 6");

    // TEST 7: Empty values for tokens
    textInput = "Hello, <name>!";
    wObject = {name: ""};
    textExpected = "Hello, !";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 7");

    // TEST 8: Repeated tokens in text
    textInput = "<name> says hello to <name>!";
    wObject = {name: "Charlie"};
    textExpected = "Charlie says hello to Charlie!";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 8");

    // TEST 9: Complex case with mixed tokens and text
    textInput = "Dear <title> <name>, welcome to the <place>.";
    wObject = {title: "Dr.", name: "Smith", place: "conference"};
    textExpected = "Dear Dr. Smith, welcome to the conference.";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 9");

    // TEST 10: Tokens with missing values should be removed
    textInput = "Hello, <name>! You live in <city>.";
    wObject = {name: "Anna"};
    textExpected = "Hello, Anna! You live in <city>.";
    textResult = stringReplaceTokens(textInput, wObject);
    assert(textExpected == textResult, "Test stringReplaceTokens_test failed: test 10");
}
```

**Comment by Alphish** [comment]
Currently, we have `string` and `string_ext` built-in functions that can replace numeric tokens with items in the given argument list/array. E.g. `var _message = string("Hello, {0}!", global.player_name);`

We also have string templates syntactic sugar where you can insert expressions directly, like so: `var _message = $"Hello, {global.player_name}!";`

*However* I do think there is value to be found in a name-based string replacement, so that you could e.g. load something like "Hello, {PLAYER}!" from one translation file or "Cześć, {PLAYER}!" from another. I'd use curly braces, because it's more common syntax for templating, both in the format strings with their {0}s, {1}s etc, and in the template strings syntactic sugar.

I'd consider a function like `string_substitute_template(_str, _substitution)` which could be used like `string_substitute_template(global.dictionary.greeting, { PLAYER: global.player_name })`.

Another possibility is to allow not only struct argument, but also a substitution function that would allow the player to process more advanced expressions. E.g. something like this:
```gml
string_substitute_template("The price of the Thorn Ring is {CURRENT_CASH+1}", { CURRENT_CASH: global.current_cash }, function(_expression, _values) {
    // _expression would be e.g. "CURRENT_CASH+1"
    // and _values would be the struct passed as the second argument
    // do the whole expression parsing with sums etc. here
});
```

I'd like to have some more discussion about what people might want from this kind of template-replacing function. I think something like e.g. customising the template delimiter (which I currently assume to be curly braces) might be better fit for a whole parsing library, but some kind of relatively generic template substitution could really come in handy!

**Comment by JujuAdams** [comment]
I'd avoid using a struct here to avoid the expensive iterations process (generating a name array from the struct, doing the struct lookup for each key). Instead, an array of key:value pairs can be used which should be a lot faster and still plenty readable.

Another point for consideration is using functions instead of static values but that is a trade-off of performance for convenience.

**Comment by Alphish** [comment]
When it comes to the implementation, I'd try to process each placeholder individually instead of mass-replacing, myself.

Reasons in favour of checking things placeholder by placeholder:
- you may be able to implement some kind of "escape", e.g. "The text \"{{PLAYER}}\" will display \"{PLAYER}\""
- the complexity increases with number of tokens only (and maybe string length in general), and not with the number of items in the struct (e.g. if you have struct of 10 values and only 2 are replaced, that's 8 replacement cycles wasted)
- the individual placeholder processing may handle irregular expressions, like "CURRENT_CASH+1", or even case-insensitive replacement

More broadly, I'd like to match the behaviour of `string` and `string_ext` templates, including how double curly braces is a de-facto escape sequence. For example:
```gml
show_debug_message(string("This is some {{0}} and this is {1}", "zero", "one"));
```
will show "This is some {0} and this is one" rather than "This is some {zero} and this is one".

**Comment by JujuAdams** [comment]
I would err on the side of simplicity. Handling irregular expressions is scope creep and, ultimately, this sort of string replacement operation should be precalculated and the result stored by the user.

**Comment by Alphish** [comment]
Just for the record, I don't mean having to try and figure out what irregular expressions users could come up with and try to make a whole system for that, but rather giving an option to inject a function that would get the result based on the expression and the struct.

It would be similar to JSON filters in that regard - regular users might not even know they exist (because the filter argument is optional), but power users can tap into it.

API-wise, it would be like this: `string_replace_template(str, values, [evaluator])`, where evaluator is a function taking struct and placeholder and returns a target value (names subject to change).

Implementation-wise, it would be a matter of replacing hardcoded struct access with the evaluator:
```gml
// find the placeholder
var _placeholder_replacement = string(_evaluator(_values, _placeholder) ?? ""); // instead of string(struct_get(_values, _placeholder) ?? "{" + _placeholder + "}"))
// add the placeholder replacement
```

I think it's this kind of tweak that opens up lots of possibilities at negligible implementation and API complexity cost (mostly thanks to the only additional argument being optional).

NOTE: The above applies if we decide to go for the individual placeholder processing implementation. If `string_replace` solution is what we're going for, then indeed supporting those kinds of filters would add some extra cost.
(though I still prefer individual placeholder processing since I don't need to worry about passing a struct with 100 or more variables, because that many happen to be relevant in the dialogue *somewhere*)

Supporting double curly brace escapes would be another bonus, but I just found that they don't actually work in the GM runtime proper (it runs a regex-replace instead - at least in HTML5). The `string("This is some {{0}} and this is {1}", "zero", "one")` thing was optimised away by C# compiler instead, which does support curly braced escapes.

**Comment by Kezarus-tech** [comment]
I'm not sold on Juju's argument about performance. This function should be used to prepare a string and should be called only once. I fail to see how this could be a performance issue on it's own. Unless the user calls it on the step/draw/alarm events or in a loop, but this is a fail of understanding on the user side, not the function. This could be said for a plethora of other built-in functions too.

And I'm all for something more generic that could handle other tokens that aren't between "<>". I used that specifically because there is not other functions or patterns that collides with it. Like "{}" or "[]" would. But I veer towards simplicity and not use a function inside it for substitution. Maybe pass it as a "StartToken" and "FinishToken" parameters?

I use this as a "fire and forget" solution to replace texts that might contain tokens that are on a **dynamic** variable bag. So I hurl the parameters there (ONCE!) to create a string that would be correct to be displayed. So the syntactic sugar solution wouldn't be good for that specific problem too. But... maybe my use case is too specific? I used that extensively on Endless RPG dialogue, with many many distinct dialogues that would have distinct variables for each Dialogue/Quest that are populated on the fly by the Dialogue/Quest engines. If that is too specific I would rather not implement this function here on the toolbox and leave each developer create their own boilerplate solution.

**Comment by Alphish** [comment]
I just remembered that "string interpolation" is a term, you can look it up here: https://en.wikipedia.org/wiki/String_interpolation

So maybe string_interpolate would be a good name. On the one hand, it may not be very intuitive for people unfamiliar with the term, on the other hand they can look it up, and once they know, it's hard to mistake for anything else.

Also, just thought of a potentially common use case for custom expression processing that doesn't have odd expressions and stuff. Namely: number formatting. E.g. for English version of the game one might want to use decimal point for fractional numbers, but for Polish version comma would be used in place of decimal points instead.

Of course, per-project needs may vary, which is why I think allowing custom interpolation/substitution function would give lots of flexibility at a little cost.

**Comment by Alphish** [comment]
Wrote an implementation of `string_interpolate` that replaces the placeholders using the given values, taking into account escapes. While the last observed behaviour for the `string` and `string_ext` functions was to ignore any escaping at runtime, [recent commits to the docs](https://github.com/YoYoGames/GameMaker-Manual/commit/22fa8dbaab7753fc80768638d6a4f97beb4fe290) in response to [this bug report](https://github.com/YoYoGames/GameMaker-Bugs/issues/8564#issuecomment-2523882358) indicate that double-brace to escape will become the intended behaviour.

More details on how string interpolation is supposed to work:
- the **placeholder** is any expression between an **unescaped opening brace** and **unescaped closing brace**, such that no brace characters are contained in the placeholder
- double opening braces are escaped into single opening braces, and double closing braces are escaped into single opening braces, e.g. `{{` will turn into `{` and `}}}}}}` will turn into `}}}`
- **unescaped opening brace** is the last opening brace in an odd-length continuous sequence of opening braces; the other braces in the sequence are treated as escaped double braces to turn into single braces
- similarly, **unescaped closing brace** is the first closing brace in an odd-length continuous sequence of closing braces; the other braces in the sequence are treated as escaped double braces to turn into single braces
- any valid **placeholder** is turned into a replacement string/value using the **evaluator function** and the values provided
- if the evaluator returns `undefined` for the given placeholder, it's assumed the placeholder cannot be replaced, and thus no replacement happens around the placeholder
- by default, the evaluator simply retrieves the replacement value from the values struct using the placeholder as the struct key

Here goes the implementation I came up with, which should meet all the conditions above. If you know some alternatives that work better, feel free to share:
```gml
function string_interpolate(_template, _values, _evaluator = struct_get) {
    // -------
    // Statics
    // -------
    
    static replace_next_placeholder = function(_context) {
        var _segments = _context.segments;
        var _count = _context.segments_count;
        var _segment_index = _context.segment_index;
        var _segment = _segments[_segment_index];
        
        // handling the closing brace
        // in particular, whether there's an even or odd number of braces to escape
        // for each pair of closing braces a single brace is added, too
        _context.segment_index++;
        var _is_escaped = false;
        
        while (_context.segment_index < _count - 1 && _segments[_context.segment_index] == "") {
            _is_escaped = !_is_escaped;
            if (_is_escaped)
                _segments[_context.segment_index] = "}";
            
            _context.segment_index++;
        }
        
        if (_is_escaped) {
            _segments[_segment_index] = string_replace_all(_segment, "{{", "{");
            return;
        }
        
        // replacing the placeholder in the current segment
        var _replaced_segment = !_is_escaped ? _context.process_segment_placeholder(_context, _segment) : undefined;
        _segments[_segment_index] = _replaced_segment ?? string_replace_all(_segment, "{{", "{") + "}";
    }
    
    static process_segment_placeholder = function(_context, _segment) {
        static trim_braces = ["{"];
        
        // could not find opening brace in the segment, not replacing the placeholder
        var _open_index = string_last_pos("{", _segment);
        if (_open_index < 1)
            return undefined;
        
        var _pre_placeholder = string_copy(_segment, 1, _open_index - 1);
        var _pre_brace = string_trim_end(_pre_placeholder, trim_braces);
        var _extra_open_braces = string_length(_pre_placeholder) - string_length(_pre_brace);
        
        // there's an even total number of opening braces (odd extra + one main)
        // the opening braces are all escaped, so not replacing the placeholder
        if (_extra_open_braces % 2 == 1)
            return undefined;
        
        var _placeholder = string_delete(_segment, 1, _open_index);
        var _result = _context.evaluator(_context.values, _placeholder);
        
        // the placeholder didn't evaluate to a valid result, not replacing the placeholder
        if (is_undefined(_result))
            return undefined;
        
        // the placeholder evaluated to a valid value, returning the replaced string
        return string_replace_all(_pre_placeholder, "{{", "{") + string(_result);
    }
    
    static context = { process_segment_placeholder: process_segment_placeholder };
    
    // -------------
    // Main function
    // -------------
    
    if (_template == "")
        return "";
    
    context.segments = string_split(_template, "}");
    context.segments_count = array_length(context.segments);
    context.segment_index = 0;
    context.values = _values;
    context.evaluator = _evaluator;
    
    // handling all the segments before the last
    while (context.segment_index < context.segments_count - 1) {
        replace_next_placeholder(context);
    }
    
    // handling the last segment
    // it's after the last brace, so it cannot have any placeholder expression
    context.segments[context.segment_index] = string_replace_all(context.segments[context.segment_index], "{{", "{");
    
    var _result = string_join_ext("", context.segments);
    context.segments = undefined;
    return _result;
}
```

I tried it out with the following code:
```gml
show_debug_message(string_interpolate("Hello, {name}!", { name: "Alice" }));
show_debug_message(string_interpolate("Hello, {{name}}!", { name: "Alice" }));
show_debug_message(string_interpolate("Hello, {{{name}}}!", { name: "Alice" }));
show_debug_message(string_interpolate("Hello, {{{{name}}}}!", { name: "Alice" }));
show_debug_message(string_interpolate("Hello, {{{{{name}}}}}!", { name: "Alice" }));

show_debug_message(string_interpolate("This is {name} and they are {age} years old!", { name: "Alice", age: 12 }));
show_debug_message(string_interpolate("This is {name} and they are {age} years old!", { name: "Bob", age: 34 }));
show_debug_message(string_interpolate("This is {name} and they are {age} years old!", { name: "Charlie" }));

show_debug_message(string_interpolate("Hello, {name}!", { name: "Alice" }, function(_values, _placeholder) { return string_upper(_values[$ _placeholder]); }));
```
and got the following output:
```
Hello, Alice!
Hello, {name}!
Hello, {Alice}!
Hello, {{name}}!
Hello, {{Alice}}!
This is Alice and they are 12 years old!
This is Bob and they are 34 years old!
This is Charlie and they are {age} years old!
Hello, ALICE!
```

**Comment by Alphish** [comment]
To have people discuss on the up-to-date proposed function (without having to scroll through the entire thread), I reformulated the suggestion in a new issue #126
